### DISCUSSION TITLE
Update consent mode after user has given consent

### DISCUSSION THREAD

**Discussion by mkue** [firebase/firebase-js-sdk#7793] [Q&A]
I'm running into an issue that, I believe, should be very common, but couldn't find an answer to (only an [unanswered Stack Overflow question](https://stackoverflow.com/questions/76347587/updating-firebase-analytics-consent-mode-for-web-app)).

I would like to use consent mode and only allow analytics to use cookies once the user has given consent through clicking on a button in a cookie consent banner. The procedure is described in this [video from Google Analytics](https://www.youtube.com/watch?v=MqAEbshMv84&t=1s).

Quick summary of the procedure:

1. Set consent settings if consent settings can be found in local storage. If local storage does contain consent settings, set all consent settings to `'denied'` and show a banner on the website that asks the user for consent. (Use `setConsent()` from Firebase SDK for this)
2. Initialize analytics (gtag) script (for GDPR-compliant tracking without cookies). (Use `getAnalytics()` or `initializeAnalytics()` from Firebase SDK)
3. Update the consent settings soon as the user gives consent and store the settings in local storage so that they are automatically loaded on the next page load. (Unclear how to do this with Firebase SDK)

The problem is that I can't find a way to update the consent settings after the `initializeAnalytics()` function has been called. According to the Google Analytics documentation, the analytics script should be initialized though, even if the user has not given any consent yet. Calling the `setConsent()` function again once the user has given consent seems to have no impact – no cookies are set in the browser after calling the function. I assume that the `gtag()` function needs to be called with the `'update'` parameter, but this isn't possible through the Firebase SDK – there is no `updateConsent()` function. Also, I could not figure out how to access or call the `gtag()` function manually after `initializeAnalytics()` has initialized the script.

The only way I have found to make Google Analytics properly work after the user has given consent is to reload the page. Then, the default consent will be based on the allowed consent stored in local storage and the `initializeAnalytics()` call will set the Google Analytics cookies. I would like to prevent this page reload though as it seems unnecessary.

Does anyone know how to update consent settings only after a user has given consent? This seems like such a common use-case to me that I assume it should be doable without a page refresh?

**Edit:**

It seems that `setConsent()` does call the `gtag` script with the `update` parameter:
https://github.com/firebase/firebase-js-sdk/blob/00235ba68fdbb5d9788a14ba2bdd75cad87301e4/packages/analytics/src/api.ts#L771

My question remains, though. How can I make sure that the analytics cookies are set in the browser and analytics events are sent when I update the consent from `denied` to `granted`?

**Comment by ivribalko** [comment]
Thank you for your message, it's been very helpful!

I can confirm that updating consent status after initialization seems impossible, neither `gtag('consent', 'update')` nor `setConsent` change event sending availability (neither enabling or disabling) set by `setConsent` before initialization for me.

>Also, I could not figure out how to access or call the gtag() function manually after initializeAnalytics() has initialized the script.

This seems to work for me, `gtag` is exported into global scope and available for calling directly.
